### ISSUE TITLE
Support for mathmlref element

### ISSUE THREAD

**Issue by robander** [dita-ot/dita-ot#2561]
Right now I'm working on support for `<mathmlref>`. This is a subset of #1774, but I don't expect to implement all of it in DITA-OT at this point so I've opened an issue specific to that element. (Once complete, whatever is done to make it work should be easily redone to make `<svgref>` work, a small subset of #1123)

The DITA specification says:
> Processors _SHOULD_ process the MathML as though the `<m:math>` element had occurred directly in the content of the containing `<mathml>` element.

http://docs.oasis-open.org/dita/dita/v1.3/os/part2-tech-content/langRef/technicalContent/mathmlref.html

I've got this part of the way done but have encountered a few issues:
1. `<mathmlref>` appears inside of a specialization of `<foreign>`, so the preprocess ignores it - the reference never registers. At this point I've set `MATHML_D_MATHMLREF` to the class value, as with other known class values, and added exceptions in `DebugFilter.java` so that it's processed. This feels like a hack. I wouldn't think this is appropriate for my own specializations. But this is an OASIS one, with an OASIS defined processing behavior, so it seems like we really should extend things to support it. For example: `if (foreignLevel <= 1 || MATHML_D_MATHMLREF.matches(atts.getValue(ATTRIBUTE_NAME_CLASS))){`
1. This feels very much like the `<coderef>` element. So I've used that debug filter to get the math targets added into `.job.xml`, because I expected a later step to pull in the math, much like with code references. But, I'm not sure this is the correct approach.
1.  Again like coderef, I tried to add a SAX filter, in the same `topic-fragment` step -- but it doesn't use the XML catalog, so couldn't parse the Math files (they have a DTD reference). I later heard from @jelovirt that the step which processes image metadata might be a good (catalog-aware) spot to add the filter, but I haven't tried this yet.
1. I'm more comfortable in XSL anyway so I tried to do that instead - but the `.job.xml` file isn't set up to retrieve only topics that use math, and throwing every DITA topic into a new XSL step is wasteful. Plus, so far, the files are processed in the source directory like code targets, and don't exist in the temp directory.

At this point I decided to seek advice. What would the ideal approach be here?

**Comment by robander** [comment]
Worth noting - the spec says that content _SHOULD_ be treated as if it was declared directly in the topic. But some transform types might actually need to export the inline math out to another file. So, _not_ resolving these references could actually be the correct approach for some transform types. That is, a transform step that actually pushes inline math into a new document might consider _not_ resolving `<mathmlref>` to be an optimization.

**Comment by drmacro** [comment]
Mathmlref should be a specialization of xref, not foreign. From the RNG:

<define name="mathmlref.attlist" combine="interleave">
      <ref name="global-atts"/>
      <optional>
        <attribute name="class" a:defaultValue="+ topic/xref mathml-d/mathmlref "/>
      </optional>
    </define>

**Comment by jelovirt** [comment]
The implementation should be similar to coderef filter. The only difference is that the it will not write to characters method, but rather forwards all events expect start and end document events.

**Comment by drmacro** [comment]
Checking the 2.4 OT it looks like mathmlref is correctly specialized from xref in all the grammars.

**Comment by robander** [comment]
> Mathmlref should be a specialization of xref, not foreign. From the RNG:

Yes -- it is -- the problem is that it's inside of `<mathml>`, which is specialized from foreign, and the DITA-OT preprocess pretty much closes its eyes when it sees a foreign element (ignores the internals, assuming it is foreign content).

**Comment by drmacro** [comment]
Note also that the referenced math may not be the only content of its containing document, so processors that publish MathML as separate resources will still need to be prepared to read referenced math from their source files in order to then copy it to the published resource--you can't assume that there is a one-to-one correspondence between references and MathML documents.

Or maybe the better way to think of it is that whether or not the math is published to a separate resource is orthogonal to whether or not it was authored inline in the DITA source or used by reference from a separate non-DITA MathML document.

**Comment by drmacro** [comment]
I see, I misread the original comment (I'll blame cedar pollen and continuing illness). I was afraid some really significant bug had eluded all of us up to now.

**Comment by jelovirt** [comment]
Processing inside the `foreign` but ignoring everything that isn't DITA (element that contains `@class` that matches the syntax of class attribute value) should be implemented as a separate refactoring PR as this MML inlining.

**Comment by robander** [comment]
I may be able to play with that but will not be quick, as mentioned I'm far more comfortable in XSLT than Java. @jelovirt sounds like the preference (if this is done properly™), would be to first refactor code so that it can handle DITA inside of `foreign` and probably `unknown`, and then take advantage of that refactoring when working on `mathmlref`?

**Comment by robander** [comment]
Is there code already that validates `@class` meets DITA's requirements? It looks like `DitaClass` assumes the proper structure and doesn't try to validate, but maybe I'm missing something.

I'm wondering at a general way to recognize `@class` attributes that match the syntax of DITA ... a simple test to catch 99% would be any `@class` that begins with (`-` or `+`) followed by white space followed by (`topic/` or `map/`). But that could theoretically be fooled into matching something _almost but not quite_ DITA.

**Comment by drmacro** [comment]
Should be able to use a regular expression that matches only complete and correct @class values, especially if the first module/name pair has to be topic/ or map/.

**Comment by shaneataylor** [comment]
@robander I have a customization I'm using to include MathML inline in HTML output when referenced with `<image>` (I started this before `<mathmlref>` existed) — I see no reason why it couldn't be adapted to this purpose.

**Comment by shaneataylor** [comment]
I also think there are synergies between implementing this correctly and implementing support for SVG (#1123).

**Comment by robander** [comment]
Out of curiosity, I tried hacking a class attribute on a normal xref so that it would appear to be `mathmlref`, just outside of the `foreign` context: `<xref format="mml" href="math-as-xref.xml" class="+ topic/xref mathml-d/mathmlref "/>`

The file appears in the `.job.xml` file, but with an HTML format: `<file src="file:/C:/dita-ot/test/math/math-as-xref.xml" uri="math-as-xref.xml" path="math-as-xref.xml" format="html">`

With that, it gets copied to the output directory as if it were an HTML target.

**Comment by robander** [comment]
@shaneataylor certainly if `<mathmlref>` is working to pull math content inline, `<svgref>` support should be nearly identical. 

For your own support - I'm guessing that by referencing the math content from an `<image>`, the math document was already copied to the temp directory for you to work with?

**Comment by jelovirt** [comment]
@robander While Saxon is fast, there is no reason why to build the source and result trees on a case like this. SAX filter is a lot more simple solution (even though there may be more line of code) and it can then be combined with other SAX filters with very little performance penalty. So 👎 for XSLT implementation.

**Comment by jelovirt** [comment]
@robander When you reference something with `image`, it's not copied to temporary directory in all cases, it depends on transtype. Also, the preprocessing code will be changed to do the copying as the last thing in the preprocessing chain, because those copy operations should not be done if there is still something that could fail and cause the process to exit.

**Comment by jelovirt** [comment]
@robander `DitaClass` currently doesn't do validation, but it will need to be added to support this use case. A regexp is a good simple solution, from a `public static boolean` method.

**Comment by robander** [comment]
@jelovirt right about the image - I was just following up on how @shaneataylor had gotten it to work by using images. And my assumption here was that SAX filter was the appropriate way to go, even if that's slower for me...

**Comment by robander** [comment]
These seem to work for testing `@class`, though I'm not sure if this is the best way to do it:

```
private static final Pattern VALID_DITA_CLASS = Pattern.compile("(\\+|-)\\s+(topic|map)/\\S+\\s+" +
                                                         "([\\S[^/]]+/\\S+\\s+)*");
```
and
```
public static boolean isValidDitaClass (final String classString) {
    	return VALID_DITA_CLASS.matcher(classString).matches();
    }
```
